### PR TITLE
health check: add Redis custom health checker

### LINKED ISSUE #2933: extensions: allow for pluggable health checkers
Would be generally useful and also needed to full cleanup/compile out Redis. For reference see https://github.com/envoyproxy/envoy/pull/2926.

### PR DESCRIPTION
**Pull request by dio**
health check: add Redis custom health checker

This PR is part of *repo-reorg* effort, to let cluster's health checker to pluggable.
The first provided custom health checker is the Redis health checker. By having this,
along with the Redis filter extension (which is a hard dep for the Redis custom health
checker), health checker can be compiled out.

The implementation of custom Redis health checker is copied from the existing 
implementation inside the `health_checker_impl.{h, cc}.`

*Risk Level*: Low, since this is only a code movement.

*Testing*:
Unit tests.

*Docs Changes*:
https://github.com/envoyproxy/data-plane-api/pull/610/commits/e0e2c401bc00cfff74e053747c77875f90218e8f

*Release Notes*:
https://github.com/envoyproxy/data-plane-api/pull/610/commits/f70dab8eb8b982915a88f42360e5c02a216bd377

Fixes #2933. 

*API Changes*:
https://github.com/envoyproxy/data-plane-api/pull/610/commits/e0e2c401bc00cfff74e053747c77875f90218e8f

Signed-off-by: Dhi Aurrahman <dio@rockybars.com>

### CODE REVIEW

**Comment by dio** [comment]
The build failure seems because of this: https://github.com/envoyproxy/envoy/issues/2851#issuecomment-378437268.

**Comment by mattklein123** [comment]
@dio can we do the PR in a single PR? I think if you copy the files GH should track it as a move and it should hopefully be pretty easy to review. I would rather not have the code living in two places at the same time. Thank you.

**Comment by dio** [comment]
OK @mattklein123, I will update here when it is ready. Especially, regarding docs.

**Comment by dio** [comment]
Really nervous about touching 20 files. But I think it is ready for a first pass. Thank you. 

cc. @mattklein123

File: test/common/upstream/health_checker_impl_test.cc [hunk 1]
```diff
@@ -65,27 +58,6 @@ envoy::api::v2::core::HealthCheck createGrpcHealthCheckConfig() {
   return health_check;
 }
 
-TEST(HealthCheckerFactoryTest, createRedis) {
-  std::string json = R"EOF(
-  {
```
Comment (danielhochman) [inline test/common/upstream/health_checker_impl_test.cc]: are there any v1 tests remaining?

Comment (dio) [reply test/common/upstream/health_checker_impl_test.cc]: I think I have moved Redis related tests into [test/extensions/health_checkers/redis/redis_test.cc](https://github.com/envoyproxy/envoy/pull/2993/files?#diff-6102697694e258451a7baf8a3a7eb745) and have it in V2.

Comment (danielhochman) [reply test/common/upstream/health_checker_impl_test.cc]: the v1 api still supported though. so we need to test it.

[cds_json.cc:54-59](https://github.com/envoyproxy/envoy/blob/master/source/common/config/cds_json.cc#L54-L59) is no longer covered: [report](https://38446-65214191-gh.circle-artifacts.com/0/build/envoy/generated/coverage/coverage.source_common_config_cds_json.cc.html)

Comment (dio) [reply test/common/upstream/health_checker_impl_test.cc]: Ah, got it. Will have it then. Thank you for mentioning it.

Comment (dio) [reply test/common/upstream/health_checker_impl_test.cc]: It is covered now. [cds_json.cc:54-59](https://39363-65214191-gh.circle-artifacts.com/0/build/envoy/generated/coverage/coverage.source_common_config_cds_json.cc.html) of cd3398e.

**Comment by dio** [comment]
@mattklein123 do you think we should put "warning is logged" as a default behaviour on deprecating a feature/config (unless it is stated not to)? Since in current deprecated list for 1.7.0, almost every item has it.

**Comment by mattklein123** [comment]
@dio thanks for your patience here. There is one more repo reorg change coming that will effect this in a small way, then we can review/get this in. I will let you know.

File: source/common/config/well_known_names.h [hunk 2]
```diff
@@ -324,5 +324,16 @@ class TransportSocketNameValues {
 
 typedef ConstSingleton<TransportSocketNameValues> TransportSocketNames;
 
+/**
```
Comment (mattklein123) [inline source/common/config/well_known_names.h]: Please see how I broke out well_known_names in https://github.com/envoyproxy/envoy/pull/3075. Can you replicate that structure when you merge master? Thank you!

Comment (dio) [reply source/common/config/well_known_names.h]: @mattklein123 OK. I have it on following commits. I hope it is fine to [require ext well-known names for a while](https://github.com/dio/envoy/blob/fc104517bc1e9d164d70490efd860f54e7d0553c/source/common/upstream/BUILD#L120) from upstream health checker (until `redis_health_check` is removed).

File: source/extensions/health_checkers/BUILD [hunk 3]
```diff
@@ -0,0 +1,18 @@
+licenses(["notice"])  # Apache 2
+# Redis custom health checker.
```
Comment (mattklein123) [inline source/extensions/health_checkers/BUILD]: Delete comment (I think just copy/paste mistake)

File: source/extensions/health_checkers/redis/config.cc [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+namespace RedisHealthChecker {
+
+Upstream::HealthCheckerSharedPtr RedisHealthCheckerFactory::createCustomHealthChecker(
+    const envoy::api::v2::core::HealthCheck& config,
+    Server::Configuration::HealthCheckerFactoryContext& context) {
+
+  return std::make_shared<RedisHealthChecker>(
+      context.cluster(), config, getRedisHealthCheckConfig(config), context.dispatcher(),
+      context.runtime(), context.random(),
+      Extensions::NetworkFilters::RedisProxy::ConnPool::ClientFactoryImpl::instance_);
```
Comment (mattklein123) [inline source/extensions/health_checkers/redis/config.cc]: nit: You can remove `Extensions::` prefix as you are already in Extensions namespace.

File: source/extensions/health_checkers/well_known_names.h [hunk 5]
```diff
@@ -0,0 +1,23 @@
+#pragma once
+
+#include "common/singleton/const_singleton.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HealthCheckers {
+
+/**
+ * Well-known access logger names.
```
Comment (mattklein123) [inline source/extensions/health_checkers/well_known_names.h]: s/access logger/health checker

File: source/extensions/health_checkers/well_known_names.h [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+ * NOTE: New access loggers should use the well known name: envoy.access_loggers.name.
```
Comment (mattklein123) [inline source/extensions/health_checkers/well_known_names.h]: envoy.health_checkers.name

File: source/extensions/health_checkers/well_known_names.h [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+ */
+class HealthCheckerNameValues {
+public:
+  // Redis health checker.
+  const std::string REDIS_HEALTH_CHECKER = "envoy.health_checker.redis";
```
Comment (mattklein123) [inline source/extensions/health_checkers/well_known_names.h]: `"envoy.health_checkers.redis"` (plural like the others)

**Comment by dio** [comment]
@mattklein123 thanks for catching the copy-paste mistakes!

**Pull request merged**